File: elements-of-programming-interviews/cpp/dutch_national_flag.cc
#include <array>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::vector;
enum class Color { kRed, kWhite, kBlue };

int pivotSort(int left, const Color& val, vector<Color>& A) {
  int curr = left;
  while (curr < A.size()) {
    if (A[curr] == val) {
      Color temp = A[curr];
      A[curr] = A[left];
      A[left] = temp;
      left++;
    }
    curr++;
  }
  return left;
}

void DutchFlagPartition(int pivot_index, vector<Color>* A_ptr) {
  /*
  This solution kind of sucks because the problem kind of sucks; as stated,
  we want to sort A into [less than pivot, equal to pivot, greater than pivot].
  However, The test cases here are based on the dutch flag, not arbitrary
  integers - so the end result is the Dutch flag no matter what, and this code
  directly solves that (with two passes) by hardcoding the given colors. This
  approach would not work for a more general Dutch Flag sort akin to quicksort's
  pivot. See my Python solution for a better approach.
  */
  vector<Color>& A = *A_ptr;
  int i = pivotSort(0, Color::kRed, A);
  pivotSort(i, Color::kWhite, A);
  return;
}

void DutchFlagPartitionWrapper(TimedExecutor& executor, const vector<int>& A,
                               int pivot_idx) {
  vector<Color> colors;
  colors.resize(A.size());
  std::array<int, 3> count = {0, 0, 0};
  for (size_t i = 0; i < A.size(); i++) {
    count[A[i]]++;
    colors[i] = static_cast<Color>(A[i]);
  }
  Color pivot = colors[pivot_idx];

  executor.Run([&] { DutchFlagPartition(pivot_idx, &colors); });

  int i = 0;
  while (i < colors.size() && colors[i] < pivot) {
    count[static_cast<int>(colors[i])]--;
    ++i;
  }

  while (i < colors.size() && colors[i] == pivot) {
    count[static_cast<int>(colors[i])]--;
    ++i;
  }

  while (i < colors.size() && colors[i] > pivot) {
    count[static_cast<int>(colors[i])]--;
    ++i;
  }

  if (i != colors.size()) {
    throw TestFailure("Not partitioned after " + std::to_string(i) +
                      "th element");
  } else if (count != std::array<int, 3>{0, 0, 0}) {
    throw TestFailure("Some elements are missing from original array");
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "A", "pivot_idx"};
  return GenericTestMain(args, "dutch_national_flag.cc",
                         "dutch_national_flag.tsv", &DutchFlagPartitionWrapper,
                         DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/python/dutch_national_flag.py
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

RED, WHITE, BLUE = range(3)


def pivot_sort(A, left, test_fn):
    curr = left
    while curr < len(A):
        if test_fn(A[curr]):
            tmp = A[left]
            A[left] = A[curr]
            A[curr] = tmp
            left += 1
        curr += 1
    return left


def dutch_flag_partition(pivot_index: int, A: List[int]) -> None:
    val = A[pivot_index]
    left = pivot_sort(A, 0, lambda x: x < val)
    pivot_sort(A, left, lambda x: x == val)
    return


@enable_executor_hook
def dutch_flag_partition_wrapper(executor, A, pivot_idx):
    count = [0, 0, 0]
    for x in A:
        count[x] += 1
    pivot = A[pivot_idx]

    executor.run(functools.partial(dutch_flag_partition, pivot_idx, A))

    i = 0
    while i < len(A) and A[i] < pivot:
        count[A[i]] -= 1
        i += 1
    while i < len(A) and A[i] == pivot:
        count[A[i]] -= 1
        i += 1
    while i < len(A) and A[i] > pivot:
        count[A[i]] -= 1
        i += 1

    if i != len(A):
        raise TestFailure('Not partitioned after {}th element'.format(i))
    elif any(count):
        raise TestFailure('Some elements are missing from original array')


if __name__ == '__main__':
    import random
    arr = [random.randint(-100, 100) for _ in range(100)]
    partition = 20
    print(f"Partitioning on {arr[partition]}")
    dutch_flag_partition(partition, arr)
    print(arr)
    """
    exit(
        generic_test.generic_test_main('dutch_national_flag.py',
                                       'dutch_national_flag.tsv',
                                       dutch_flag_partition_wrapper))
    """
